spec/cspec/c/export-kernel-builds.py

```diff
@@ -0,0 +1,180 @@
+# and optimisation level. With L4V_ARCH=RISCV64 and L4V_FEATURES=MCS, for
+# example:
+
+# RISCV64-MCS-O1/
+# RISCV64-MCS-O1/kernel_all.cpp
+# RISCV64-MCS-O1/kernel.elf
+# RISCV64-MCS-O1/...
+# RISCV64-MCS-O2/
+# RISCV64-MCS-O2/kernel_all.cpp
+# RISCV64-MCS-O2/kernel.elf
+# RISCV64-MCS-O1/...
+
+import argparse
+import os
+import shutil
+import subprocess
+import sys
+
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import NamedTuple, Optional
+
+
+class L4vPaths(NamedTuple):
+    kernel_mk: Path
+    c_pp: Path
+    c_functions: Path
+
+
+def get_l4v_paths(l4v_arch: str) -> L4vPaths:
+    l4v_c = Path(__file__).resolve().parent
+    l4v = l4v_c.parent.parent.parent
+    kernel_mk = l4v_c / 'kernel.mk'
+    c_pp = l4v_c / f'build/{l4v_arch}/kernel_all.c_pp'
+    c_functions = l4v / f'proof/asmrefine/export/{l4v_arch}/CFunDump.txt'
+    return L4vPaths(kernel_mk=kernel_mk, c_pp=c_pp, c_functions=c_functions)
+
+
+class ExportConfig(NamedTuple):
+    export_root: Path
+    l4v_arch: str
+    l4v_features: Optional[str]
+    l4v_paths: L4vPaths
+    manifest: Optional[Path]
+
+    def config_name(self, optimisation: str) -> str:
+        features = f'-{self.l4v_features}' if self.l4v_features else ''
+        return f'{self.l4v_arch}{features}{optimisation}'
+
+    def export_path(self, optimisation: str) -> Path:
+        return self.export_root / self.config_name(optimisation)
+
+    def do_export(self, optimisation: str):
+        config_name = self.config_name(optimisation)
+        export_dir = self.export_path(optimisation)
+
+        print(f'Exporting kernel build for {config_name} to {export_dir}...')
+
+        if not self.l4v_paths.c_pp.is_file():
+            print('  Note: No l4v kernel build found.')
+        if not self.l4v_paths.c_functions.is_file():
+            print('  Note: No C graph-lang found.')
+
+        with TemporaryDirectory() as build_tmp:
+            env = {
+                **os.environ,
+                'KERNEL_BUILD_ROOT': Path(build_tmp) / 'build',
+                'KERNEL_EXPORT_DIR': export_dir,
+                'CONFIG_OPTIMISATION': optimisation
+            }
+            p = subprocess.run(['make', '-f', self.l4v_paths.kernel_mk, 'kernel_export'],
+                               env=env, stdin=subprocess.DEVNULL)
+        if p.returncode != 0:
+            print(
+                f'Error: Kernel build for {config_name} failed', file=sys.stderr)
+            sys.exit(1)
+
+        # Copy the preprocessed source from the l4v build, so we can later compare
+        # it to the exported builds if binary verification fails. There's no point
+        # comparing here, because there will always be differences due to lines
+        # inserted by the preprocessor and config system.
+        if self.l4v_paths.c_pp.is_file():
+            shutil.copyfile(self.l4v_paths.c_pp, export_dir /
+                            'kernel_all.c_pp.l4v')
```

Comment:
I think @Xaphiosis has a way of stripping those lines if that helps.

Reply:
Yeah, that would be cool.

Reply:
I might try to figure this out later, but I'll leave it as it is for now.